(#452) Ver.E Visibility Adjustment When Upgrading a Friend → Close Friend

tyzhou05 opened the pull request:
## Issue Number: #452

## Self Check List

- [x] Have you double-checked the base branch into which this PR will be merged?
- [x] Have you rebased from the base branch?

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (formatting, renaming)
- [ ] Merge Develop to Main
- [ ] Other (please describe):

## What does this PR do?

### **Version E Implementation (Connection Updates)**
* Added ability to _upgrade_ or _downgrade_ connection status between users (friend → close_friend; close_friend → friend) 
* Implemented visibility adjustment when upgrading friendship level; users must choose to apply the visibility change _only to future posts_ or to _all past posts_ as well 
  * Added past posts visibility control through `update_past_posts` flag

### Version E Implementation Details
* Modified ConnectionChoiceUpdate view to use `GenericAPIView`, with choices: ['`friend`', '`close_friend`']
Added custom `get_object()` method to find connections via friend's ID
* Endpoints: 
  * PUT/PATCH `/api/user/connections/<int:pk>/` - Update connection level; can upgrade or downgrade specific users to `friend` or `close_friend` and decide whether they want to show their past posts with `upgrade_past_posts` to `true` or `false` accordingly .
      * JSON body input example:
        * {
       "choice": "close_friend",
       "update_past_posts": false
      } 
    * [close_friends](https://diivers.postman.co/workspace/WhoAmI~c2317c3a-2ab1-458e-96b5-01f8ca6b983e/request/6673035-af546cf8-e358-43a8-b0a1-c8e74c7df241?tab=body)
    * [friends](https://diivers.postman.co/workspace/WhoAmI~c2317c3a-2ab1-458e-96b5-01f8ca6b983e/request/6673035-7b9862da-120f-472a-8a5a-1c39893ba9d9?tab=body)
  * While testing, view friends / close_friends and their updated connection statuses via `/api/user/friends/?type=all` :
    * [all_friends](https://diivers.postman.co/workspace/WhoAmI~c2317c3a-2ab1-458e-96b5-01f8ca6b983e/request/6673035-13790cf4-d9ba-44f4-86c5-85ad28e5da30)

## Preview Image

## Further comments

tyzhou05 committed:
initial changes in Connection model

njs03332 commented on adoorback/qna/migrations/0009_alter_response_visibility.py:
```diff
@@ -0,0 +1,18 @@
+# Generated by Django 4.2.14 on 2025-02-02 23:44
```
I don't think this migration file is necessary since it doesn't indicate any changes since the last migration file! (adoorback/qna/migrations/0008_response_visibility.py)

tyzhou05 replied:
Got it, removed!

njs03332 commented on adoorback/account/models.py:
```diff
@@ -447,6 +452,62 @@ def user2_is_close_friend(self):
+                    visibility='friends',
+                    created_at__lt=current_time  # Only update posts created before upgrade
+                )
+            notes.update(visibility='close_friends')
```
The visibility of the posts themselves must not change. The spec is as follows:

Let's say User A changes Friend B from *friend* to *close friend*.
- If `update_past_posts` is **true**, all of A's previous posts that has visibility *close friends* will now be visible to B.
- If `update_past_posts` is **false**, A's previous *close friends* posts will **not** be visible to B, and only future posts shared with *close friends* will be visible.

To implement this feature, you'll likely need to modify the `is_audience` function.  
When checking if a user is part of the audience, if the post's visibility is set to *close friends*, you need to do more than just verify if the user is a *close friend*.  
(If `update_past_posts` is **false**, you will also need to check if the post's `created_at` timestamp is **after** the time when the user was upgraded to *close friend* (`userx_upgrade_time`).  )

Let me know if anything is unclear!

tyzhou05 replied:
I see, thanks for re-explaining! I've removed most of what I initially did and changed my implementation to match the spec:
* Removed the code that was updating post visibilities (notes.update(visibility='close_friends'))
* Changed to use UpdateAPIView with PATCH method instead of POST
* Kept the existing `ConnectionChoiceUpdate` view and just modified it
* Checked for visibility within `is_audience` as suggested

I think this should work since it doesn't change the post visibilities - instead, we're just controlling what B can see through the is_audience function based on:
* When they were upgraded to close friend
* Whether update_past_posts is true/false
* Changes in `Response` qna/models and `Note` account/models

I hope this makes sense and is clearer! Let me know if my understanding makes sense :)

njs03332 commented on adoorback/account/urls.py:
```diff
@@ -42,7 +42,9 @@
     path('friends/hidden/', views.UserHiddenAdd.as_view(), name='user-hidden-add'),
     path('friends/<int:pk>/hidden/', views.UserHiddenDestroy.as_view(), name='user-hidden-destroy'),
 
-    path('connections/<int:pk>/', views.ConnectionChoiceUpdate.as_view(), name='connection-choice-update'),
```
I think we could slightly modify the existing `ConnectionChoiceUpdate` view to achieve the new functionality. Was there any particular reason it was difficult to do so?

Also, using PUT or PATCH instead of POST seems more appropriate since we are just modifying some fields of a model, though it might be due to my limited understanding. Could you explain this part as well?

tyzhou05 replied:
Ah I see! I've removed the changes I made to this API, and slightly edited the `ConnectionChoiceUpdate` view instead. To check my understanding, what are the main differences between using PUT/PATCH versus POST - is it because we are only updating fields and do not need to create new ones?

njs03332 replied:
Yes, exactly! As far as I know, POST is used to create new resources while PUT/PATCH is used to update an existing resource.

njs03332 commented on adoorback/account/views.py:
```diff
@@ -1119,3 +1120,18 @@ def list(self, request, *args, **kwargs):
         return Response(serialized_data)
+
+
+class ConnectionViewSet(viewsets.ModelViewSet):
```
If there's no issue, I think it would be more consistent and readable to use something like `UpdateAPIView` from `generics` (or another appropriate view) instead of `ModelViewSet`, in line with our existing practices!

tyzhou05 replied:
Yes, that's my bad! I renamed it to `ConnectionChoiceUpdate` and updated it to use `generics.UpdateAPIView`, do you think that makes more sense?

njs03332 replied:
Yes! I think you can remove this class since we're not using it anymore!

njs03332 reviewed (changes requested):
@tyzhou05 I left a few comments :) Let me know if you need more explanation!

tyzhou05 committed:
addressing comments + changing implementation

tyzhou05 commented:
@njs03332 Just made some changes and addressed the comments in the threads. I think there are definitely some things to round out -- thanks in advance for going back and forth with me!

njs03332 commented on adoorback/qna/models.py:
```diff
@@ -145,8 +145,28 @@ def is_audience(self, user):
             return True
 
+        # Get connection between author and user
+        connection = Connection.objects.filter(
```
You can also just use the pre-defined `get_connection_between` classmethod here! (decined in account/models.py)

njs03332 commented on adoorback/qna/models.py:
```diff
@@ -145,8 +145,28 @@ def is_audience(self, user):
+                    return is_close and self.created_at > connection.user2_upgrade_time
+                return is_close
         
         return self.author.is_connected(user)
```
I think we can just return `True` here since we already checked if connection exists at line 153.

njs03332 commented on adoorback/qna/models.py:
```diff
@@ -145,8 +145,28 @@ def is_audience(self, user):
+
         if self.visibility == 'close_friends':
-            return self.author.is_close_friend(user)
+            # Check if user is a close friend
```
Here you can use `is_close_friend()` method defined under the `User` model in `account/models.py`!
(It’s better this way because if any Connection/User related code changes in the future, you’ll only need to update the method.)

njs03332 commented on adoorback/note/models.py:
```diff
@@ -106,8 +107,28 @@ def is_audience(self, user):
             print("is_audience: author is user")
             return True
 
+        # Get connection between author and user
```
Same here !

njs03332 commented on adoorback/account/views.py:
```diff
@@ -767,20 +768,18 @@ def perform_destroy(self, obj):
-    serializer_class = ConnectionChoiceUpdateSerializer
     permission_classes = [IsAuthenticated]
-
-    def get_object(self):
```
I think you shouldn't remove the `get_object()` method here!

If you call `get_object()` (like you did in the `patch` method) without overriding it, it will try to find the `Connection` object using the `pk` value from the URL. However, the frontend can't know the `pk` of the `Connection` between the current user and their friend in advance.

Instead, the frontend knows which friend the current user wants to change the status for. So, it can send the friend's `pk`, and we will find the corresponding `Connection` object using that `pk` (as I did in the previous `get_object()` method).

tyzhou05 replied:
Got it! 

Just to confirm what I've just changed: I've kept the `get_object() `method since the frontend needs to use the friend's `pk` to find the correct Connection. I've added serializer validation to the `patch()` method while maintaining its existing structure. And here, we still need `get_object() `because the frontend knows the friend's `pk` but not the Connection `pk` in advance, correct?

njs03332 replied:
Hmm I can't find the serializer validation you mentioned you added to the patch method..! 

Yes you're right! If we don't override `get_object()` then it tries to find the object using `pk` from the `queryset` defined at line 769 ([source code](https://github.com/encode/django-rest-framework/blob/28d0261afcd6702900512e00c37f4e264c117d83/rest_framework/generics.py#L79)). 
We will receive the friend's `pk` (not the connection's `pk`) from the frontend so we need to find the `Connection` object _through_ this friend. Therefore we need to override the `get_object()` method here!

tyzhou05 replied:
Oh I'm so sorry that's my misunderstanding again! I thought I had overriden the serializer with `patch()` but I think the new API view works without one now -- thanks for clarifying again over why overriding makes sense!

njs03332 commented on adoorback/account/views.py:
```diff
@@ -767,20 +768,18 @@ def perform_destroy(self, obj):
 
 class ConnectionChoiceUpdate(generics.UpdateAPIView):
     queryset = Connection.objects.all()
-    serializer_class = ConnectionChoiceUpdateSerializer
```
I think the way you implemented it will work perfectly, but it seems like modifying the existing serializer could have made the process a bit easier. (Just something to keep in mind for next time!)

Also, if we're not planning to use this serializer, it might be a good idea to remove it from serializers.py as well!

tyzhou05 replied:
Sounds good, removed and will keep that in mind!

tyzhou05 replied:
Actually, after I removed this serializer I'm now running into an error about this while testing. I'm not sure if my implementation is right now or not -- could you take a look at this?
![Screenshot 2025-02-04 at 9 29 02 PM](https://github.com/user-attachments/assets/e8fa1810-860f-47ba-90d4-ca919e00d957)

njs03332 replied:
Ah I think defining a serializer is necessary in `UpdateAPIView`. If you don't want to define a serializer, you can use `GenericAPIView` instead! (reference: `MarkAllNotificationsRead` in `notification/views.py`)

cf) Our general standard practice is to use serializers when validating and serializing data. (Using serializers makes it easier to perform data validation, improves code consistency, and enhances reusability.) However, the current codebase doesn't strictly follow this practice! :) So, while the way you've implemented it is totally fine, it might be better to use serializers next time!

tyzhou05 replied:
Ahh I see. I changed it to `GenericAPIView` and the Postman tests are working without a serializer now! Will keep that in mind for the future, thanks for the clarification :)

njs03332 reviewed (commented):
LGTM except for some refactoring I suggested in the comments!
After applying my comments and testing with Postman, it’d be great if you could document the API specs clearly for the frontend team to review, just like last time.

Thanks for being so quick :)

tyzhou05 committed:
addressing comments, starting to test now

tyzhou05 commented:
Thanks for the comments @njs03332! Besides the comment thread about removing the `ConnectionChoiceUpdateSerializer,` everything makes sense to me except that part -- I might be a little confused about what removing it does. 

A summary of what I changed:
* Removed serializer_class = `ConnectionChoiceUpdateSerializer` from the view since we're handling validation manually 
* Removed the ConnectionChoiceUpdateSerializer class from serializers.py as it's no longer being used
* Kept the get_object() validation in the patch() method
* Removed the `ConnectionViewSet` view

njs03332 commented on adoorback/qna/models.py:
```diff
@@ -145,10 +145,25 @@ def is_audience(self, user):
         if self.visibility == 'close_friends':
-            return self.author.is_close_friend(user)
+            is_close = self.author.is_close_friend(user)
+            if connection.user1 == user:
```
I don't think you need branching by `connection.user1` here, since you already retrieved the value of `is_close` at line 154

njs03332 commented on adoorback/note/models.py:
```diff
@@ -106,8 +107,22 @@ def is_audience(self, user):
+
         if self.visibility == 'close_friends':
-            return self.author.is_close_friend(user)
+            if connection.user1 == user:
```
Here as well!

tyzhou05 committed:
Updated to GenericAPIView; refactoring redundant code

tyzhou05 commented:
@njs03332 Thanks for the review -- just addressed everything and updated the documentation with the Postman tests. Let me know if there's anything to add to the documentation or if everything looks good to merge or not!

njs03332 reviewed (approved):
Why is the http method (put/patch) for `/api/user/connections/<int:pk>/` set differently depending on the "choice" field?
Also, when registering the API in Postman, you only need to add one entry per URL. There's no need to register multiple entries based on different field values—since the frontend team can manually change the field values for testing.

Everything looks good to merge! Just check my comments related to Postman. :) Great work!

tyzhou05 commented:
> Why is the http method (put/patch) for `/api/user/connections/<int:pk>/` set differently depending on the "choice" field? Also, when registering the API in Postman, you only need to add one entry per URL. There's no need to register multiple entries based on different field values—since the frontend team can manually change the field values for testing.
> 
> Everything looks good to merge! Just check my comments related to Postman. :) Great work!

Oops that's my bad; I thought it would make more sense to test `friend` and `true` / `close_friend` and `false`. But now I do see it makes more sense to focus on one (`close_friend` and `true`/`false`) for the tests and let frontend change the values on their own! That makes more sense given your explanation.

Thanks for reviewing! Merging into main now

tyzhou05 committed:
Merge branch 'main' into 452-visibility

njs03332 commented on adoorback/account/views.py:
```diff
@@ -1180,17 +1192,3 @@ def list(self, request, *args, **kwargs):
         return Response(serialized_data)
-
-
-class CurrentUserVersionChange(APIView):
```
Current main branch is throwing this error
```python
  File "/Users/yuri/Documents/git/WhoAmI-Today-backend/adoorback/account/urls.py", line 26, in <module>
    path('me/version-change/', views.CurrentUserVersionChange.as_view(), name='current-user-version-change'),
AttributeError: module 'account.views' has no attribute 'CurrentUserVersionChange'
```
This is something I implemented [here](https://github.com/ssm0318/WhoamI-Today-backend/pull/445), but it looks like the related code might have been removed when you rebased with main and resolved merge conflicts.

Could you check your merge process again and ensure that no code gets lost?

tyzhou05 replied:
Ahh hmm I thought I had rebased properly before I merged, I'm so sorry it seems that I deleted that! Should I just recreate the branch for this issue and re-merge the class in directly? So sorry once again!

njs03332 replied:
That's alright it's hard not to make mistakes while using git :). 
I just reverted the merge to the main branch and restored your branch. You could go back to the commit before the rebase on the 452-visibility branch, rebase again, and merge it again into the main branch.

tyzhou05 replied:
Thanks! I've re-merged the branch into main and restored the `CurrentUserVersionChange`, let me know if everything looks good now

njs03332 replied:
Yes everything looks good to me :) Appreciate your effort!
